identity/src/keypair.rs

```diff
@@ -224,42 +218,73 @@ impl Keypair {
             #[cfg(all(feature = "rsa", not(target_arch = "wasm32")))]
-            Self::Rsa(_) => return Err(DecodingError::encoding_unsupported("RSA")),
+            Self::Rsa(data) => proto::PrivateKey {
+                Type: proto::KeyType::RSA,
+                Data: data.to_raw_bytes(),
+            },
             #[cfg(feature = "secp256k1")]
-            Self::Secp256k1(_) => return Err(DecodingError::encoding_unsupported("secp256k1")),
+            Self::Secp256k1(data) => proto::PrivateKey {
+                Type: proto::KeyType::Secp256k1,
+                Data: data.secret().to_bytes().into(),
+            },
             #[cfg(feature = "ecdsa")]
-            Self::Ecdsa(_) => return Err(DecodingError::encoding_unsupported("ECDSA")),
+            Self::Ecdsa(data) => proto::PrivateKey {
+                Type: proto::KeyType::ECDSA,
+                Data: data.secret().to_bytes(),
+            },
         };
 
         let mut buf = Vec::with_capacity(pk.get_size());
         let mut writer = Writer::new(&mut buf);
         pk.write_message(&mut writer).expect("Encoding to succeed");
 
-        Ok(buf)
+        buf
     }
 
     /// Decode a private key from a protobuf structure and parse it as a [`Keypair`].
-    #[cfg_attr(not(feature = "ed25519"), allow(unused_mut))]
+    #[deprecated(
+        since = "0.2.0",
+        note = "This method name is inaccurate, use `Keypair::try_from_protobuf_encoding` instead."
+    )]
     pub fn from_protobuf_encoding(bytes: &[u8]) -> Result<Keypair, DecodingError> {
+        Self::try_from_protobuf_encoding(bytes)
+    }
+
+    /// Try to decode a private key from a protobuf structure and parse it as a [`Keypair`].
+    #[cfg_attr(not(feature = "ed25519"), allow(unused_mut))]
+    pub fn try_from_protobuf_encoding(bytes: &[u8]) -> Result<Keypair, DecodingError> {
         use quick_protobuf::MessageRead;
 
         let mut reader = BytesReader::from_bytes(bytes);
         let mut private_key = proto::PrivateKey::from_reader(&mut reader, bytes)
             .map_err(|e| DecodingError::bad_protobuf("private key bytes", e))
             .map(zeroize::Zeroizing::new)?;
 
+        #[allow(deprecated, unreachable_code)]
         match private_key.Type {
-            #[cfg(feature = "ed25519")]
-            proto::KeyType::Ed25519 =>
-            {
-                #[allow(deprecated)]
-                ed25519::Keypair::decode(&mut private_key.Data).map(Keypair::Ed25519)
+            proto::KeyType::Ed25519 => {
+                #[cfg(feature = "ed25519")]
+                return ed25519::Keypair::try_decode(&mut private_key.Data).map(Keypair::Ed25519);
+                Err(DecodingError::missing_feature("ed25519"))
+            }
+            proto::KeyType::RSA => {
+                #[cfg(all(feature = "rsa", not(target_arch = "wasm32")))]
+                return rsa::Keypair::try_decode_pkcs8(&mut private_key.Data)
+                    .or(rsa::Keypair::try_decode_der(&mut private_key.Data))
```

Comment:
Why are we attempting to decode from `der` here? The spec says it will be encoded in pkcs8. If it is not inside a pkcs8, we should fail parsing it.